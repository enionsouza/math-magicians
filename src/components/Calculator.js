import { useState } from 'react';
import Button from './Button';
import calculate from '../logic/calculate';
import './Calculator.css';

const Calculator = () => {
  const [state, setState] = useState(
    {
      total: null,
      next: null,
    },
  );

  const onClickHandler = (e) => {
    const { total, next } = state;
    if (!total && !next && e.target.classList.contains('operator')) return null;
    setState((state) => calculate(state, e.target.innerHTML));
    return null;
  };

  const utilLabels = ['AC', '+/-', '%'];
  const digitLabels = ['7', '8', '9', '4', '5', '6', '1', '2', '3', '0', '.'];
  const operatorLabels = ['รท', 'x', '-', '+', '='];
  const { total, next } = state;

  const utils = utilLabels.map(
    (utilLabel, index) => (
      <Button
        key={`${utilLabel}`}
        cssStyle={{ gridArea: `util-${index}` }}
        cssClass="util"
        action={onClickHandler}
        label={utilLabel}
      />
    ),
  );
  const digits = digitLabels.map(
    (digitLabel, index) => (
      <Button
        key={`${digitLabel}`}
        cssStyle={{ gridArea: `digit-${index}` }}
        cssClass="digit"
        action={onClickHandler}
        label={digitLabel}
      />
    ),
  );
  const operators = operatorLabels.map(
    (operatorLabel, index) => (
      <Button
        key={`${operatorLabel}`}
        cssStyle={{ gridArea: `operator-${index}` }}
        cssClass="operator"
        action={onClickHandler}
        label={operatorLabel}
      />
    ),
  );
  return (
    <div className="grid">
      <div style={{ gridArea: 'display' }} className="display">{ next || total || '0' }</div>
      {utils}
      {digits}
      {operators}
    </div>
  );
};

export default Calculator;
