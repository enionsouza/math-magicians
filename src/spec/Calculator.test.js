import { render, fireEvent, screen } from '@testing-library/react';
import Calculator from '../components/Calculator';

describe('Calculator component', () => {
  it('Calculator snapshot test', () => {
    const { asFragment: calculator } = render(<Calculator />);

    expect(calculator()).toMatchSnapshot();
  });

  describe('Calculator interaction tests', () => {
    it('test interaction: 50.1 + 50 = 100.1, then press AC', () => {
      const { container } = render(<Calculator />);
      const display = container.querySelector('.display');

      ['5', '0', '.', '1'].forEach((input) => {
        fireEvent.click(screen.getByText(input));
      });

      expect(display.innerHTML).toEqual('50.1');

      ['+', '5', '0', '='].forEach((input) => {
        fireEvent.click(screen.getByText(input));
      });

      expect(display.innerHTML).toEqual('100.1');

      fireEvent.click(screen.getByText('AC'));

      expect(display.innerHTML).toEqual('0');
    });
  });
});
