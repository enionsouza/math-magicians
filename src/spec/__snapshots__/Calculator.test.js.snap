// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator component Calculator snapshot test 1`] = `
<DocumentFragment>
  <div
    class="grid"
  >
    <div
      class="display"
      style="grid-area: display;"
    >
      0
    </div>
    <button
      class="util"
      style="grid-area: util-0;"
      type="button"
    >
      AC
    </button>
    <button
      class="util"
      style="grid-area: util-1;"
      type="button"
    >
      +/-
    </button>
    <button
      class="util"
      style="grid-area: util-2;"
      type="button"
    >
      %
    </button>
    <button
      class="digit"
      style="grid-area: digit-0;"
      type="button"
    >
      7
    </button>
    <button
      class="digit"
      style="grid-area: digit-1;"
      type="button"
    >
      8
    </button>
    <button
      class="digit"
      style="grid-area: digit-2;"
      type="button"
    >
      9
    </button>
    <button
      class="digit"
      style="grid-area: digit-3;"
      type="button"
    >
      4
    </button>
    <button
      class="digit"
      style="grid-area: digit-4;"
      type="button"
    >
      5
    </button>
    <button
      class="digit"
      style="grid-area: digit-5;"
      type="button"
    >
      6
    </button>
    <button
      class="digit"
      style="grid-area: digit-6;"
      type="button"
    >
      1
    </button>
    <button
      class="digit"
      style="grid-area: digit-7;"
      type="button"
    >
      2
    </button>
    <button
      class="digit"
      style="grid-area: digit-8;"
      type="button"
    >
      3
    </button>
    <button
      class="digit"
      style="grid-area: digit-9;"
      type="button"
    >
      0
    </button>
    <button
      class="digit"
      style="grid-area: digit-10;"
      type="button"
    >
      .
    </button>
    <button
      class="operator"
      style="grid-area: operator-0;"
      type="button"
    >
      รท
    </button>
    <button
      class="operator"
      style="grid-area: operator-1;"
      type="button"
    >
      x
    </button>
    <button
      class="operator"
      style="grid-area: operator-2;"
      type="button"
    >
      -
    </button>
    <button
      class="operator"
      style="grid-area: operator-3;"
      type="button"
    >
      +
    </button>
    <button
      class="operator"
      style="grid-area: operator-4;"
      type="button"
    >
      =
    </button>
  </div>
</DocumentFragment>
`;
